File: day51/threeSum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new LinkedList<>();
        for (int i = 0; i < nums.length; i++) {
            if (i > 0 && nums[i] == nums[i - 1] || nums[i] > 0) continue;
            int j = i + 1, k = nums.length - 1;
            while (j < k) {
                int theSum = nums[i] + nums[j] + nums[k];
                if (theSum < 0) j++;
                else if (theSum > 0) k--;
                else {
                    res.add(Arrays.asList(nums[i], nums[j], nums[k]));
                    j++;
                    k--;
                    while (j < k && nums[j] == nums[j - 1]) j++;
                    while (j < k && nums[k] == nums[k + 1]) k--;
                }
            }
        }
        return res;
    }
}

// class Solution {
//     public List<List<Integer>> threeSum(int[] nums) {
//         Arrays.sort(nums);
//         Set<List<Integer>> res = new LinkedHashSet<>();
//         for (int i = 0; i < nums.length - 2; i++) {
//             if (i > 0 && nums[i] == nums[i - 1] || nums[i] > 0) continue;
//             Set<Integer> dic = new HashSet<>(nums.length - i);
//             for (int j = i + 1; j < nums.length; j++) {
//                 Integer tmp = - nums[i] - nums[j];
//                 if (dic.contains(nums[j])) {
//                     List<Integer> list = Arrays.asList(nums[i], tmp, nums[j]);
//                     res.add(list);
//                 } else {
//                     dic.add(tmp);
//                 }
//             }
//         }
//         return new ArrayList<>(res);
//     }
// }

// class Solution {
//     public List<List<Integer>> threeSum(int[] nums) {
//         Set<List<Integer>> result = new LinkedHashSet<>();

//         for (int i = 0; i < nums.length - 2; i++) {
//             int target = -nums[i];
//             Map<Integer, Integer> hashMap = new HashMap<>(nums.length - i);
//             for (int j = i + 1; j < nums.length; j++) {
//                 int v = target - nums[j];
//                 Integer exist = hashMap.get(v);
//                 if (exist != null) {
//                     List<Integer> list = Arrays.asList(nums[i], exist, nums[j]);
//                     list.sort(Comparator.naturalOrder());
//                     result.add(list);
//                 } else {
//                     hashMap.put(nums[j], nums[j]);
//                 }
//             }
//         }

//         return new ArrayList<>(result);
//     }
// }

File: day51/threeSum.py
/*
15. 三数之和
给你一个包含 n 个整数的数组 nums，判断 nums 中是否存在三个元素 a，b，c ，使得 a + b + c = 0 ？请你找出所有满足条件且不重复的三元组。

注意：答案中不可以包含重复的三元组。
https://leetcode-cn.com/problems/3sum/
 */

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if len(nums) < 3: return []
        nums.sort()
        res = set()
        for i, v in enumerate(nums[:-2]):
            if i >= 1 and v == nums[i - 1] or v > 0: continue
            d = set()
            for x in nums[i + 1:]:
                if x not in d:
                    d.add(-v-x)
                else:
                    res.add((v, -v-x, x))
        
        return list(map(list, res))

# class Solution:
#     def threeSum(self, nums: List[int]) -> List[List[int]]:
#         res = []
#         nums.sort()
#         for i in range(len(nums) - 2):
#             if i > 0 and nums[i] == nums[i - 1] or nums[i] > 0: continue
#             j, k = i + 1, len(nums) - 1
#             while j < k:
#                 three_sum = nums[i] + nums[j] + nums[k]
#                 if three_sum > 0:
#                     k -= 1
#                 elif three_sum < 0:
#                     j += 1
#                 elif three_sum == 0:
#                     res.append([nums[i], nums[j], nums[k]])
#                     k -= 1
#                     j += 1
#                     while nums[k] == nums[k + 1] and j < k: k -= 1
#                     while nums[j] == nums[j - 1] and j < k: j += 1
#         return res
